src/dataflow/src/storage/render/upsert.rs

```diff
@@ -552,35 +552,43 @@ where
 
 /// `thin` uses information from the source description to find which indexes in the row
 /// are keys and skip them.
-fn thin(key_indices: &[usize], value: &Row, row_buf: &mut Row) -> Row {
+fn thin(key_indices: &BTreeMap<usize, usize>, value: &Row, row_buf: &mut Row) -> Row {
     let mut row_packer = row_buf.packer();
     let values = &mut value.iter();
-    let mut next_idx = 0;
-    for &key_idx in key_indices {
-        // First, push the datums that are before `key_idx`
-        row_packer.extend(values.take(key_idx - next_idx));
-        // Then, skip this key datum
-        values.next().unwrap();
-        next_idx = key_idx + 1;
+    for (value_idx, value) in values.enumerate() {
+        if !key_indices.contains_key(&value_idx) {
+            row_packer.push(value)
+        }
     }
-    // Finally, push any columns after the last key index
-    row_packer.extend(values);
-
     row_buf.clone()
 }
 
 /// `rehydrate` uses information from the source description to find which indexes in the row
```

Comment:
```suggestion
/// `rehydrate` uses information from the source description to find which indices in the row
```

might as well be consistent here